service/controller/resource/alerting/alertmanagerconfig/create.go

```diff
@@ -0,0 +1,59 @@
+package alertmanagerconfig
+
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	desired, err := r.toSecret()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.Debugf(ctx, "creating")
+	current, err := r.k8sClient.K8sClient().CoreV1().Secrets(desired.GetNamespace()).Get(ctx, desired.GetName(), metav1.GetOptions{})
+	if apierrors.IsNotFound(err) {
+		current, err = r.k8sClient.K8sClient().CoreV1().Secrets(desired.GetNamespace()).Create(ctx, desired, metav1.CreateOptions{})
+	}
+
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if r.hasChanged(current, desired) {
+		updateMeta(current, desired)
+		_, err = r.k8sClient.K8sClient().CoreV1().Secrets(desired.GetNamespace()).Update(ctx, desired, metav1.UpdateOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+	r.logger.Debugf(ctx, "created")
+
+	return nil
+}
+
+func updateMeta(c, d metav1.Object) {
+	d.SetGenerateName(c.GetGenerateName())
+	d.SetUID(c.GetUID())
+	d.SetResourceVersion(c.GetResourceVersion())
+	d.SetGeneration(c.GetGeneration())
+	d.SetSelfLink(c.GetSelfLink())
+	d.SetCreationTimestamp(c.GetCreationTimestamp())
+	d.SetDeletionTimestamp(c.GetDeletionTimestamp())
+	d.SetDeletionGracePeriodSeconds(c.GetDeletionGracePeriodSeconds())
+	// without this, it's impossible to change labels on resources
+	if len(d.GetLabels()) == 0 {
+		d.SetLabels(c.GetLabels())
+	}
+	// without this, it's impossible to change annotations on resources
+	if len(d.GetAnnotations()) == 0 {
+		d.SetAnnotations(c.GetAnnotations())
+	}
+	d.SetFinalizers(c.GetFinalizers())
+	d.SetOwnerReferences(c.GetOwnerReferences())
+	d.SetManagedFields(c.GetManagedFields())
+}
```

Comment:
This would be a simpler way of updating the object's metadata, but I also thing we should not include such refactoring in here as there's already a lot going on.

```suggestion
		desired.ObjectMeta = current.ObjectMetada
		_, err = r.k8sClient.K8sClient().CoreV1().Secrets(desired.GetNamespace()).Update(ctx, desired, metav1.UpdateOptions{})
		if err != nil {
			return microerror.Mask(err)
		}
	}
	r.logger.Debugf(ctx, "created")

	return nil
}
```

Reply:
That's not simpler, that's breaking. Because this does not accept label or annotation changes

Reply:
Yeah a shared package would be nice.

Reply:
It is added here https://github.com/giantswarm/prometheus-meta-operator/pull/1641/files#diff-f9121ba3abf019bf78f83931b583dd84ad4cf19d50fecca517ba6eed29cfde7b :)